### PR TITLE
Fix early stopping on_training_epoch end with structured results

### LINKED ISSUE #3413: Early stopping does not work with structured result returned in training_epoch_end
## 🐛 Bug

Situation: A structured result is returned in ``training_epoch_end`` and ``early_stop_on`` was set.
Illustrative example:
```
    def training_epoch_end(self, outputs):
        avg_loss = outputs['minimize'].mean().squeeze()
        result = pl.TrainResult(early_stop_on=avg_loss)
        return result
```

Expected Behaviour: early stopping will be triggered
Actual Behaviour: it is not triggered

### To Reproduce

Run the code sample below.

#### Code sample

Try running my example: https://gist.github.com/Lucas-Steinmann/f900b3ef5636028d5917d4cb0183a291

### Expected behavior

Should early stop after about 20 epochs, since metric given to ``early_stop_on`` is artificially increased.

### Environment

* CUDA:
        - GPU:
                - GeForce GTX 1080 Ti
        - available:         True
        - version:           10.1
* Packages:
        - numpy:             1.19.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.6.0
        - pytorch-lightning: 0.9.1rc1
        - tensorboard:       2.2.0
        - tqdm:              4.48.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.3
        - version:           #116~16.04.1-Ubuntu SMP Wed Aug 26 17:36:48 UTC 2020

### Additional context

First discussed here: https://github.com/PyTorchLightning/pytorch-lightning/pull/3193#discussion_r484581575

### PR DESCRIPTION
**Pull request by Lucas-Steinmann**
## What does this PR do?

First, adds a test which tries to set ``early_stop_on`` in a ``TrainResult`` in ``training_epoch_end`` and tests if early stopping is triggered.
The test fails.

Second, fixes the issue by storing the ``callback_metrics`` for structured results in ``logger_connecter.py``.

Third, fixes a bug in a test model, which returns a tensor with more than one element for ``checkpoint_on``.
This bug was hidden before, since ``checkpoint_on`` was also discarded alongside ``early_stop_on``.

Fixes  #3413

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes? --> Will be done in #3193
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by williamFalcon** [comment]
mind waiting till next week until we are done with refactors?
anything results related is likely to change by then, so i don't want you to waste time if we overwrite these changes

**Comment by Lucas-Steinmann** [comment]
> mind waiting till next week until we are done with refactors?

no problem. I'm not really depending on the feature anyway.

**Comment by edenlightning** [comment]
@williamFalcon i think most refactors are done and this can be synced, right?

**Comment by edenlightning** [comment]
We changed our API from structured results to a simpler API, [please look at the latest docs](https://pytorch-lightning.readthedocs.io/en/latest/loggers.html). We now pass early stoping as a callback. please try to update to master, see if issue remains!

**Comment by edenlightning** [comment]
I'm closing this PR since this should no longer supported, please reopen if needed!
